consensus/proto_array/src/proto_array_fork_choice.rs

```diff
@@ -262,6 +279,73 @@ impl ProtoArrayForkChoice {
             .map_err(|e| format!("find_head failed: {:?}", e))
     }
 
+    pub fn get_proposer_head<E: EthSpec>(
+        &self,
+        justified_state_balances: &[u64],
+        canonical_head: Hash256,
+        re_org_vote_fraction: u64,
+    ) -> Result<ProposerHead, String> {
+        let nodes = self
+            .proto_array
+            .iter_nodes(&canonical_head)
+            .take(2)
+            .collect::<Vec<_>>();
+        if nodes.len() != 2 {
+            return Ok(ProposerHead::default());
+        }
+        let head_node = nodes[0];
+        let parent_node = nodes[1];
+
+        // Re-org conditions.
+        let is_single_slot_re_org = parent_node.slot + 1 == head_node.slot;
+        let re_org_weight_threshold =
+            calculate_proposer_boost::<E>(justified_state_balances, re_org_vote_fraction)
+                .ok_or_else(|| {
+                    "overflow calculating committee weight for proposer boost".to_string()
+                })?;
+        let canonical_head_weight = self
+            .get_block_unique_weight(canonical_head, justified_state_balances)
+            .map_err(|e| format!("overflow calculating head weight: {:?}", e))?;
+        let is_weak_head = canonical_head_weight < re_org_weight_threshold;
+
+        let re_org_head = (is_single_slot_re_org && is_weak_head).then(|| parent_node.root);
```

Comment:
Note to self: still need to add the condition that we shouldn't re-org if `(head_node.slot + 1) % E::slots_per_epoch() == 0` from my prev comment https://github.com/sigp/lighthouse/pull/2860#issuecomment-1049411525.